document.go

```diff
@@ -186,3 +241,49 @@ func (d *Document) Error() string {
 	}
 	return errStr
 }
+
+/*
+MarshalJSON handles the custom serialization case caused by case where the "data"
+element of a document might be either a single resource object, or a collection of
+them.
+*/
+func (d *Document) MarshalJSON() ([]byte, error) {
+	type MarshalDoc Document
+	doc := MarshalDoc(*d)
+
+	// get raw data so we can perform top level operations
+	rawData, err := json.Marshal(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	// map back to top level JSON map for ease of use, SHITTY, I KNOW.
+	jDoc := map[string]interface{}{}
```

Comment:
How about creating a subtype of `Document` for each serialization mode? These types can live completely within `MarshalJSON`:
```go
func (d *Document) MarshalJSON() ([]byte, error) {
    type MarshalDoc Document

    doc := MarshalDoc(*d)
    switch d.Mode {
    case ObjectMode:
        var data *Object
        if len(d.Data) > 0 {
            data = d.Data[0]
        }
        // subtype that overrides data with a single object
        type MarshalObject struct {
            MarshalDoc
            Data *Object `json:"data"`
        }

        return json.Marshal(MarshalObject{
            MarshalDoc: doc,
            Data:       data,
        })
    case ErrorMode:
        // subtype that omits data
        type MarshalError struct {
            MarshalDoc
            Data *Object `json:"data,omitempty"`
        }

        return json.Marshal(MarshalError{
            MarshalDoc: doc,
        })
    }
    return json.Marshal(doc)
}
```